### PR TITLE
design draft

### PR DESCRIPTION
**Pull request by brumar**
Only a design proposition at the moment.

Relates to https://github.com/souzatharsis/podcastfy/issues/21

This design philosophy aims low at a low level modularity (3 kinds of backend passed at the init, with simple ABCs).
The podcast building is seen as a multi-staged approach, but with rewind and rebuild capabilities.

If the design is approved, it should be feasible to refactor existing code so that: 

- generate_podcast should not change its expected arguments (very straigthforward to use, very good to attract users) but will work behind the scene with the creation of three backends implementing the required methods and will call finalize to produce the podcast instance.

So user would just have to call podcast.save("./filepath.mp3).

I am absolutely open to feedback of course!

### CODE REVIEW

File: podcastfy/core/podcast.py [hunk 1]
```diff
@@ -1,64 +1,17 @@
...
-            if 'youtube.com' in parsed_url.netloc or 'youtu.be' in parsed_url.netloc:
-                return self.parsers['youtube'].parse(url)
-            return self.parsers[parsed_url.scheme].parse(url)
-        elif url.lower().endswith('.pdf'):
-            return self.parsers['pdf'].parse(url)
-        else:
-            raise ValueError(f"Unsupported URL format: {url}")
+    TRANSCRIPT_BUILT = 1
+    AUDIO_SEGMENTS_BUILT = 2
+    ASSEMBLED = 3
```
Comment (souzatharsis) [inline podcastfy/core/podcast.py:14]: Consider adding in-line comments to explain what each static value represents, e.g. What's 'ASSEMBLED'?

File: podcastfy/core/podcast.py [hunk 2]
```diff
@@ -141,25 +94,22 @@ def wrapper(self, *args, **kwargs):
 
 
 class Podcast:
```
Comment (souzatharsis) [inline podcastfy/core/podcast.py:96]: So the idea is that a lot of the "assembling" logic in podcastfy/client.py generate_podcast() and process_links() would be replace by Podcast(), is that right?

File: podcastfy/core/podcast.py [hunk 3]
```diff
@@ -141,25 +94,22 @@ def wrapper(self, *args, **kwargs):
...
-    def __init__(self, urls: List[str], llm_backend: LLMBackend, tts_backend: TTSBackend,
-                 parsing_backend: ParserCollection):
-        self.urls = urls
+    def __init__(self, content: str, llm_backend: LLMBackend, tts_backend: TTSBackend):
```
Comment (souzatharsis) [inline podcastfy/core/podcast.py:97]: There should be a Class Transcript and PodcastAudio, and Podcast should contain both. So its init would have

self.transcript = Transcript()
self.audio = PodcastAudio()

in self.transcript the user would be able to access attributes such as transcript language, style, length etc
in self.audio the user would be able to access attributes such as audio language, length, format etc

What do you think?

Comment (brumar) [reply podcastfy/core/podcast.py:97]: Just to be sure I understand, these attributes would then be used to generate the text or audio content ?
If yes, that was not really the idea behind this design which is quite low level and generic. To get back to the style things, they would be used to build the tts_backend I think. But other classes would help to make this instanciation easier. Maybe if you agree with the overall design, I should try soon to bridge your code with mine to see if it works ?

But here, the user car certainly edit the transcript or the audio segments then proceed to the next build stage.

I think I should have self.transcript = None and self.audio_segments=None and self.assembled_audio=None at the init time though.

assembled is maybe not a great name, it's the final audio in contrast with audio_segments, so maybe just audio is better.

**Comment by souzatharsis** [comment]
Should we close this one and consolidate it under #61 ?

**Comment by brumar** [comment]
Sure!

**Pull request closed without merging**